demux/utils/novaseq_samplesheet.py

```diff
@@ -0,0 +1,181 @@
+
+            return self
+
+    def remove_unwanted_indexes(self):
+        """ Filter out indexes of unwanted length and single indexes """
+
+        if self.indexlength:
+            if self.pad and int(self.indexlength) in (16, 20):
+                self.raw_samplesheet = [
+                    line
+                    for line in self.raw_samplesheet
+                    if len(line["index"].replace("-", ""))
+                    in (16, int(self.indexlength))
+                ]
+            else:
+                self.raw_samplesheet = [
+                    line
+                    for line in self.raw_samplesheet
+                    if len(line["index"].replace("-", "")) == int(self.indexlength)
+                ]
+        else:
+            self.raw_samplesheet = [
+                line for line in self.raw_samplesheet if "-" in line["index"]
+            ]
+
+        return self
+
+    def adapt_indexes(self):
+        """Adapts the indexes: pads all indexes so that all indexes have a length equal to the
+        number  of index reads, and takes the reverse complement of index 2 in case of the new
+        novaseq software control version (1.7) in combination with the new reagent kit
+        (version 1.5)"""
+
+        is_reverse_complement = (
+            self.runparameters.control_software_version == "1.7.0"
+            and self.reagent_kit_version() == 1.5
+        )
+
+        for line in self.raw_samplesheet:
+            index1, index2 = line["index"].split("-")
+            if self.pad and len(index1) == 8:
+                line["index"], line["index2"] = self.pad_and_rc_indexes(
+                    index1, index2, is_reverse_complement
+                )
+            elif len(index2) == 10:
+                line["index2"] = (
+                    self.reverse_complement(index2) if is_reverse_complement else index2
+                )
+            else:
+                line["index"], line["index2"] = index1, index2
+
+        return self
+
+    def pad_and_rc_indexes(self, index1, index2, rev_comp):
+        """ Pads and reverse complements indexes """
+
+        if self.runparameters.index_reads == 8:
+            index2 = self.reverse_complement(index2) if rev_comp else index2
+        if self.runparameters.index_reads == 10:
+            index1 = index1 + "AT"
+            index2 = (
+                self.reverse_complement("AC" + index2) if rev_comp else index2 + "AC"
+            )
+
+        return index1, index2
+
+    def reagent_kit_version(self):
+        """ Derives the reagent kit version from the run parameters """
+
+        parameter_to_version = {"1": 1.0, "3": 1.5}
+
+        return parameter_to_version.get(self.runparameters.reagent_kit_version)
+
+    def construct_samplesheet(self, end="\n", delimiter=","):
+        """ Construct the sample sheet """
+
+        demux_samplesheet = [delimiter.join(self.header)]
+        self.add_dummy_indexes().remove_unwanted_indexes().adapt_indexes()
```

Comment:
Should this not be split on different lines? Each with a corresponding unit test

Reply:
I will still add unit tests for all three.

I though [method chainging](https://code.activestate.com/recipes/580616-python-method-chaining-examples/) like this would look clean and would eliminate the need for storing intermediate results. It's something @patrikgrenfeldt showed me.

Reply:
As long as the subs are really small and "pure" i.e no sideeffects, I can see the benefit